src/Compiler/Optimize/LowerComputedCollections.fs

```diff
@@ -252,15 +252,77 @@ let (|SeqToArray|_|) g expr =
     | ValApp g g.seq_to_array_vref (_, [seqExpr], m) -> Some (seqExpr, m)
     | _ -> None
 
+[<return: Struct>]
+let (|ConstInt32Range|_|) g expr =
+    match expr with
+    | ValApp g g.range_int32_op_vref ([], [Expr.Const (value = Const.Int32 start); Expr.Const (value = Const.Int32 1); Expr.Const (value = Const.Int32 finish)], _), _ -> ValueSome (start, finish)
+    | _ -> ValueNone
+
+[<return: Struct>]
+let (|Int32Range|_|) g expr =
+    match expr with
+    | ValApp g g.range_int32_op_vref ([], [start; Expr.Const (value = Const.Int32 1); finish], _), _ -> ValueSome (start, finish)
+    | _ -> ValueNone
+
 let LowerComputedListOrArrayExpr tcVal (g: TcGlobals) amap overallExpr =
     // If ListCollector is in FSharp.Core then this optimization kicks in
     if g.ListCollector_tcr.CanDeref then
+        let constListSizeThreshold = 100
+        let constArraySizeThreshold = int System.UInt16.MaxValue
 
         match overallExpr with
+        // [5..1] → []
+        | SeqToList g (OptionalCoerce (OptionalSeq g amap (ConstInt32Range g (start, finish))), m) when
+            start > finish
+            ->
+            Some (mkUnionCaseExpr (g.nil_ucref, [g.int32_ty], [], m))
+
+        // [1..5] → [1; 2; 3; 4; 5] ≡ 1 :: 2 :: 3 :: 4 :: 5 :: []
+        | SeqToList g (OptionalCoerce (OptionalSeq g amap (ConstInt32Range g (start, finish))), _) when
+            finish - start < constListSizeThreshold
+            ->
+            let rec conses acc n =
```

Comment:
consts?

Reply:
As in multiple list [cons](https://en.wikipedia.org/wiki/Cons)es ` … :: … :: …`, since that's what it's building up.

Reply:
Ohhh okay.

Reply:
dcc970d